nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py

```diff
@@ -0,0 +1,500 @@
+        batch["original_size_as_tuple"] = torch.tensor(
+            [self.width, self.height], device=torch.cuda.current_device()
+        ).repeat(batch_size, 1)
+        batch["target_size_as_tuple"] = torch.tensor(
+            [self.width, self.height], device=torch.cuda.current_device()
+        ).repeat(batch_size, 1)
+        batch["crop_coords_top_left"] = torch.tensor([0, 0], device=torch.cuda.current_device()).repeat(batch_size, 1)
+        return batch
+
+    @torch.no_grad()
+    def generate_log_images(self, latents, batch, model):
+
+        with torch.cuda.amp.autocast(
+            enabled=self.autocast_dtype in (torch.half, torch.bfloat16), dtype=self.autocast_dtype,
+        ):
+            # get sampler (contains discretizer, scaler, guider)
+            # params = self.cfg.sampling.base
+            # sampler = get_sampler_config(params)
+            sampler = self.sampler
+
+            batch_c = self.append_sdxl_size_keys(batch)
+            batch_size = len(batch)
+
+            force_uc_zero_embeddings = ["txt", "captions"]  # force zero embeddings for text and captions
+            cond, u_cond = model.conditioner.get_unconditional_conditioning(
+                batch_c, batch_uc=None, force_uc_zero_embeddings=force_uc_zero_embeddings,
+            )
+            additional_model_inputs = {}  # not necessary for now
+
+            denoiser = lambda input, sigma, c: model.denoiser(model.model, input, sigma, c, **additional_model_inputs)
+            samples_z = sampler(denoiser, latents, cond=cond, uc=u_cond)
+            samples_x = model.decode_first_stage(samples_z)
+            vae_decoder_output = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0) * 255.0
+
+            log_reward = [
+                self.reward_model.get_reward(
+                    vae_decoder_output[i].unsqueeze(0).detach().permute(0, 2, 3, 1), batch  # (C, H, W) -> (1, H, W, C)
+                ).item()
+                for i in range(batch_size)
+            ]
+            log_img = [
+                vae_decoder_output[i].float().detach().permute(1, 2, 0).cpu().numpy()  # (C, H, W) -> (H, W, C)
+                for i in range(batch_size)
+            ]
+            return log_img, log_reward, vae_decoder_output
+
+    @torch.no_grad()
+    def log_visualization(self, prompts):
+        batch_size = len(prompts)
+        # Get different seeds for different dp rank
+        with get_cuda_rng_tracker().fork(get_data_parallel_rng_tracker_name()):
+            latents = torch.randn(
+                [
+                    batch_size,
+                    self.in_channels,
+                    self.height // self.downsampling_factor,
+                    self.width // self.downsampling_factor,
+                ],
+                generator=None,
+            ).to(torch.cuda.current_device())
+
+        image_draft_p, reward_draft_p, vae_decoder_output_draft_p = self.generate_log_images(
+            latents + 0, prompts, self.model
```

Comment:
it just creates a new copy of the tensor in case its modified by some in-place op inside the function. 

I guess I could also use `.clone()` function, that'd be cleaner.

Reply:
Yea, I like your proposal of using .clone()